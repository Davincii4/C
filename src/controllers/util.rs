use super::prelude::*;
use crate::util::errors::{forbidden, internal, AppError, AppResult};
use conduit_router::RequestParams;

/// The Origin header (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin)
/// is sent with CORS requests and POST requests, and indicates where the request comes from.
/// We don't want to accept authenticated requests that originated from other sites, so this
/// function returns an error if the Origin header doesn't match what we expect "this site" to
/// be: https://crates.io in production, or http://localhost:port/ in development.
pub fn verify_origin(req: &dyn RequestExt) -> AppResult<()> {
    let headers = req.headers();
    let allowed_origins = &req.app().config.allowed_origins;

    let bad_origin = headers
        .get_all(header::ORIGIN)
        .iter()
        .find(|value| !allowed_origins.contains(value));

    if let Some(bad_origin) = bad_origin {
        let error_message =
            format!("only same-origin requests can be authenticated. got {bad_origin:?}");
        return Err(internal(&error_message).chain(forbidden()));
    }
    Ok(())
}

pub trait RequestParamExt<'a> {
    fn param(self, key: &str) -> Option<&'a str>;
}

impl<'a> RequestParamExt<'a> for &'a (dyn RequestExt + 'a) {
    fn param(self, key: &str) -> Option<&'a str> {
        self.params().find(key)
    }
}
