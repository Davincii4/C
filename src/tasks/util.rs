pub(super) mod advisory_lock;
