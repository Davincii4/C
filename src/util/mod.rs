pub use self::errors::{bad_request, human, internal, internal_error, CargoError, CargoResult};
pub use self::errors::{std_error, ChainError};
pub use self::io_util::{read_fill, LimitErrorReader, read_le_u32};
pub use self::json::{json_error, json_response, json_error_200};
pub use self::request_proxy::RequestProxy;

pub mod errors;
pub mod rfc3339;
mod io_util;
mod json;
mod request_proxy;
