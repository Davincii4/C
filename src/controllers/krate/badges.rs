//! Endpoints that provide badges based on crate metadata

use crate::controllers::frontend_prelude::*;

use crate::models::{Badge, Crate, CrateBadge, MaintenanceStatus};
use crate::schema::*;

use conduit::{Body, Response};

/// Handles the `GET /crates/:crate_id/maintenance.svg` route.
pub fn maintenance(req: &mut dyn RequestExt) -> EndpointResult {
    let name = &req.params()["crate_id"];
    let conn = req.db_read_only()?;

    let krate = Crate::by_name(name).first::<Crate>(&*conn);
    if krate.is_err() {
        let response = Response::builder().status(404).body(Body::empty()).unwrap();

        return Ok(response);
    }

    let krate = krate.unwrap();

    let maintenance_badge: Option<CrateBadge> = CrateBadge::belonging_to(&krate)
        .select((badges::crate_id, badges::all_columns))
        .load::<CrateBadge>(&*conn)?
        .into_iter()
        .find(|cb| matches!(cb.badge, Badge::Maintenance { .. }));

    let status = maintenance_badge
        .map(|it| match it.badge {
            Badge::Maintenance { status } => Some(status),
            _ => None,
        })
        .flatten();

    let badge = generate_badge(status);

    let response = Response::builder()
        .status(200)
        .body(Body::from_vec(badge.into_bytes()))
        .unwrap();

    Ok(response)
}

fn generate_badge(status: Option<MaintenanceStatus>) -> String {
    let message = match status {
        Some(MaintenanceStatus::ActivelyDeveloped) => "actively-developed",
        Some(MaintenanceStatus::PassivelyMaintained) => "passively-maintained",
        Some(MaintenanceStatus::AsIs) => "as-is",
        Some(MaintenanceStatus::None) => "unknown",
        Some(MaintenanceStatus::Experimental) => "experimental",
        Some(MaintenanceStatus::LookingForMaintainer) => "looking-for-maintainer",
        Some(MaintenanceStatus::Deprecated) => "deprecated",
        None => "unknown",
    };

    let color = match status {
        Some(MaintenanceStatus::ActivelyDeveloped) => "brightgreen",
        Some(MaintenanceStatus::PassivelyMaintained) => "yellowgreen",
        Some(MaintenanceStatus::AsIs) => "yellow",
        Some(MaintenanceStatus::None) => "lightgrey",
        Some(MaintenanceStatus::Experimental) => "blue",
        Some(MaintenanceStatus::LookingForMaintainer) => "orange",
        Some(MaintenanceStatus::Deprecated) => "red",
        None => "lightgrey",
    };

    let badge_options = badge::BadgeOptions {
        subject: "maintenance".to_owned(),
        status: message.to_owned(),
        color: color.to_string(),
    };

    badge::Badge::new(badge_options).unwrap().to_svg()
}
