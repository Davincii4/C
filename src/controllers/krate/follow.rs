//! Endpoints for managing a per user list of followed crates

use crate::auth::AuthCheck;
use diesel::associations::Identifiable;

use crate::controllers::frontend_prelude::*;
use crate::db::DieselPooledConn;
use crate::models::{Crate, Follow};
use crate::schema::*;

fn follow_target(
    req: &ConduitRequest,
    conn: &DieselPooledConn<'_>,
    user_id: i32,
) -> AppResult<Follow> {
    let crate_name = req.param("crate_id").unwrap();
    let crate_id = Crate::by_name(crate_name)
        .select(crates::id)
        .first(&**conn)?;
    Ok(Follow { user_id, crate_id })
}

/// Handles the `PUT /crates/:crate_id/follow` route.
pub fn follow(req: ConduitRequest) -> AppResult<Response> {
    let user_id = AuthCheck::default().check(&req)?.user_id();
    let conn = req.app().db_write()?;
    let follow = follow_target(&req, &conn, user_id)?;
    diesel::insert_into(follows::table)
        .values(&follow)
        .on_conflict_do_nothing()
        .execute(&*conn)?;

    ok_true()
}

/// Handles the `DELETE /crates/:crate_id/follow` route.
pub fn unfollow(req: ConduitRequest) -> AppResult<Response> {
    let user_id = AuthCheck::default().check(&req)?.user_id();
    let conn = req.app().db_write()?;
    let follow = follow_target(&req, &conn, user_id)?;
    diesel::delete(&follow).execute(&*conn)?;

    ok_true()
}

/// Handles the `GET /crates/:crate_id/following` route.
pub fn following(req: ConduitRequest) -> AppResult<Json<Value>> {
    use diesel::dsl::exists;

    let user_id = AuthCheck::only_cookie().check(&req)?.user_id();
    let conn = req.app().db_read_prefer_primary()?;
    let follow = follow_target(&req, &conn, user_id)?;
    let following =
        diesel::select(exists(follows::table.find(follow.id()))).get_result::<bool>(&*conn)?;

    Ok(Json(json!({ "following": following })))
}
