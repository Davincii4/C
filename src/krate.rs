use std::io::MemWriter;
use std::collections::hashmap::{HashMap, Occupied, Vacant};
use std::sync::Arc;
use serialize::json;
use serialize::hex::ToHex;
use time::Timespec;

use conduit::{Request, Response};
use conduit_router::RequestParams;
use curl::http;
use pg::types::ToSql;
use pg::{PostgresRow, PostgresStatement};
use semver;

use app::{App, RequestApp};
use db::{Connection, RequestTransaction};
use dependency::Dependency;
use git;
use user::User;
use util::{RequestUtils, CargoResult, Require, internal, ChainError, human};
use util::{LimitErrorReader, HashingReader};
use util::errors::{NotFound, CargoError};
use version::{Version, EncodableVersion};

#[deriving(Clone)]
pub struct Crate {
    pub id: i32,
    pub name: String,
    pub user_id: i32,
    pub updated_at: Timespec,
    pub created_at: Timespec,
    pub downloads: i32,
    pub max_version: semver::Version,
}

#[deriving(Encodable, Decodable)]
pub struct EncodableCrate {
    pub id: String,
    pub name: String,
    pub versions: Vec<i32>,
    pub updated_at: String,
    pub created_at: String,
    pub downloads: i32,
    pub max_version: String,
}

impl Crate {
    pub fn from_row(row: &PostgresRow) -> Crate {
        let max: String = row.get("max_version");
        Crate {
            id: row.get("id"),
            name: row.get("name"),
            user_id: row.get("user_id"),
            updated_at: row.get("updated_at"),
            created_at: row.get("created_at"),
            downloads: row.get("downloads"),
            max_version: semver::Version::parse(max.as_slice()).unwrap(),
        }
    }

    pub fn find(conn: &Connection, id: i32) -> CargoResult<Crate> {
        let stmt = try!(conn.prepare("SELECT * FROM crates \
                                      WHERE id = $1"));
        match try!(stmt.query(&[&id])).next() {
            Some(row) => Ok(Crate::from_row(&row)),
            None => Err(NotFound.box_error()),
        }
    }

    pub fn find_by_name(conn: &Connection, name: &str) -> CargoResult<Crate> {
        let stmt = try!(conn.prepare("SELECT * FROM crates \
                                      WHERE name = $1 LIMIT 1"));
        match try!(stmt.query(&[&name as &ToSql])).next() {
            Some(row) => Ok(Crate::from_row(&row)),
            None => Err(NotFound.box_error()),
        }
    }

    pub fn find_or_insert(conn: &Connection, name: &str,
                          user_id: i32) -> CargoResult<Crate> {
        // TODO: like with users, this is sadly racy

        let stmt = try!(conn.prepare("SELECT * FROM crates WHERE name = $1"));
        let mut rows = try!(stmt.query(&[&name as &ToSql]));
        match rows.next() {
            Some(row) => return Ok(Crate::from_row(&row)),
            None => {}
        }
        let stmt = try!(conn.prepare("INSERT INTO crates \
                                      (name, user_id, created_at,
                                       updated_at, downloads, max_version) \
                                      VALUES ($1, $2, $3, $4, 0, '0.0.0') \
                                      RETURNING *"));
        let now = ::now();
        let mut rows = try!(stmt.query(&[&name as &ToSql, &user_id, &now, &now]));
        Ok(Crate::from_row(&try!(rows.next().require(|| {
            internal("no crate returned")
        }))))
    }

    pub fn valid_name(name: &str) -> bool {
        if name.len() == 0 { return false }
        name.chars().all(|c| c.is_alphanumeric() || c == '_' || c == '-')
    }

    fn encodable(self, versions: Vec<i32>) -> EncodableCrate {
        let Crate { name, created_at, updated_at, downloads,
                      max_version, .. } = self;
        EncodableCrate {
            id: name.clone(),
            name: name,
            versions: versions,
            updated_at: ::encode_time(updated_at),
            created_at: ::encode_time(created_at),
            downloads: downloads,
            max_version: max_version.to_string(),
        }
    }

    pub fn versions(&self, conn: &Connection) -> CargoResult<Vec<Version>> {
        let stmt = try!(conn.prepare("SELECT * FROM versions \
                                      WHERE crate_id = $1"));
        let rows = try!(stmt.query(&[&self.id]));
        Ok(rows.map(|r| Version::from_row(&r)).collect())
    }

    pub fn dl_path(&self, version: &str) -> String {
        format!("/crates/{}/{}/download", self.name, version)
    }

    pub fn s3_path(&self, version: &str) -> String {
        format!("/pkg/{}/{}-{}.tar.gz", self.name, self.name, version)
    }

    pub fn encode_many(conn: &Connection, crates: Vec<Crate>)
                       -> CargoResult<Vec<EncodableCrate>> {
        if crates.is_empty() {
            return Ok(vec![]);
        }

        let trans = try!(conn.transaction());

        try!(trans.execute("CREATE TEMPORARY TABLE crateids (
                                id INT PRIMARY KEY
                            ) ON COMMIT DROP", []));

        let mut map = {
            let mut query = MemWriter::new();
            let _ = write!(query, "INSERT INTO crateids (id) VALUES ");
            let mut crateids: Vec<&ToSql> = vec![];
            let mut first = true;
            for (i, krate) in crates.iter().enumerate() {
                if !first {
                    let _ = write!(query, ", ");
                }
                first = false;
                let _ = write!(query, "(${})", i+1);
                crateids.push(&krate.id);
            }
            let query = String::from_utf8(query.unwrap()).unwrap();
            try!(trans.execute(query.as_slice(), crateids.as_slice()));

            let stmt = try!(conn.prepare("SELECT v.id, v.crate_id FROM versions v
                                          INNER JOIN crateids c ON v.id = c.id"));

            let mut map = HashMap::new();
            for row in try!(stmt.query(&[])) {
                match map.entry(row.get("crate_id")) {
                    Occupied(e) => e.into_mut(),
                    Vacant(e) => e.set(Vec::new()),
                }.push(row.get("id"));
            }
            map
        };

        try!(trans.finish());

        Ok(crates.into_iter().map(|p| {
            let id = p.id;
            p.encodable(map.pop(&id).unwrap())
        }).collect())
    }

    pub fn add_version(&self, conn: &Connection, num: &str)
                       -> CargoResult<Version> {
        let ver = semver::Version::parse(num).unwrap();
        let max = if ver > self.max_version {&ver} else {&self.max_version};
        let max = max.to_string();
        try!(conn.execute("UPDATE crates SET updated_at = $1, max_version = $2
                           WHERE id = $3",
                          &[&::now(), &max, &self.id]));
        Version::insert(conn, self.id, num)
    }
}

pub fn index(req: &mut Request) -> CargoResult<Response> {
    let conn = try!(req.tx());
    let query = req.query();
    let page = query.find_equiv(&"page").map(|s| s.as_slice())
                    .and_then(from_str::<i64>).unwrap_or(1);
    let limit = query.find_equiv(&"per_page").map(|s| s.as_slice())
                     .and_then(from_str::<i64>).unwrap_or(10);
    if limit > 100 { return Err(human("cannot request more than 100 crates")) }
    let offset = (page - 1) * limit;
    let pattern = query.find_equiv(&"letter")
                       .map(|s| s.as_slice().char_at(0).to_lowercase())
                       .map(|s| format!("{}%", s))
                       .unwrap_or("%".to_string());

    // Collect all the crates
    let stmt = try!(conn.prepare("SELECT * FROM crates \
                                  WHERE name LIKE $3 \
                                  LIMIT $1 OFFSET $2"));
    let mut crates = Vec::new();
    for row in try!(stmt.query(&[&limit, &offset, &pattern])) {
        crates.push(Crate::from_row(&row));
    }
    let crates = try!(Crate::encode_many(conn, crates));

    // Query for the total count of crates
    let stmt = try!(conn.prepare("SELECT COUNT(*) FROM crates \
                                  WHERE name LIKE $1"));
    let row = try!(stmt.query(&[&pattern])).next().unwrap();
    let total = row.get(0u);

    #[deriving(Encodable)]
    struct R { crates: Vec<EncodableCrate>, meta: Meta }
    #[deriving(Encodable)]
    struct Meta { total: i64 }

    Ok(req.json(&R {
        crates: crates,
        meta: Meta { total: total },
    }))
}

pub fn summary(req: &mut Request) -> CargoResult<Response> {
    let tx = try!(req.tx());
    let num_crates = {
        let stmt = try!(tx.prepare("SELECT COUNT(*) FROM crates"));
        let mut rows = try!(stmt.query(&[]));
        rows.next().unwrap().get("count")
    };
    let num_downloads = {
        let stmt = try!(tx.prepare("SELECT total_downloads FROM metadata"));
        let mut rows = try!(stmt.query(&[]));
        rows.next().unwrap().get("total_downloads")
    };

    let to_crates = |stmt: PostgresStatement| {
        let rows = try!(stmt.query([]));
        Crate::encode_many(tx, rows.map(|r| Crate::from_row(&r)).collect())
    };
    let new_crates = try!(tx.prepare("SELECT * FROM crates \
                                        ORDER BY created_at DESC LIMIT 10"));
    let just_updated = try!(tx.prepare("SELECT * FROM crates \
                                        ORDER BY updated_at DESC LIMIT 10"));
    let most_downloaded = try!(tx.prepare("SELECT * FROM crates \
                                           ORDER BY downloads DESC LIMIT 10"));

    #[deriving(Encodable)]
    struct R {
        num_downloads: i64,
        num_crates: i64,
        new_crates: Vec<EncodableCrate>,
        most_downloaded: Vec<EncodableCrate>,
        just_updated: Vec<EncodableCrate>,
    }
    Ok(req.json(&R {
        num_downloads: num_downloads,
        num_crates: num_crates,
        new_crates: try!(to_crates(new_crates)),
        most_downloaded: try!(to_crates(most_downloaded)),
        just_updated: try!(to_crates(just_updated)),
    }))
}

pub fn show(req: &mut Request) -> CargoResult<Response> {
    let name = &req.params()["crate_id"];
    let conn = try!(req.tx());
    let krate = try!(Crate::find_by_name(&*conn, name.as_slice()));
    let versions = try!(krate.versions(&*conn));

    #[deriving(Encodable)]
    struct R { krate: EncodableCrate, versions: Vec<EncodableVersion>, }
    Ok(req.json(&R {
        krate: krate.clone().encodable(versions.iter().map(|v| v.id).collect()),
        versions: versions.into_iter().map(|v| v.encodable(&krate)).collect(),
    }))
}
#[deriving(Encodable)]
pub struct NewCrate {
    pub name: String,
    pub vers: String,
    pub deps: Vec<Dependency>,
    pub cksum: String,
    pub features: HashMap<String, Vec<String>>,
}

pub fn new(req: &mut Request) -> CargoResult<Response> {
    let app = req.app().clone();

    let (mut new_crate, user) = try!(parse_new_headers(req));

    // Persist the new crate, if it doesn't already exist
    let krate = try!(Crate::find_or_insert(try!(req.tx()),
                                           new_crate.name.as_slice(),
                                           user.id));
    if krate.user_id != user.id {
        return Err(human("crate name has already been claimed by another user"))
    }

    // Persist the new version of this crate
    match try!(Version::find_by_num(try!(req.tx()), krate.id,
                                    new_crate.vers.as_slice())) {
        Some(..) => {
            return Err(human(format!("crate version `{}` is already uploaded",
                                     new_crate.vers)))
        }
        None => {}
    }
    let vers = try!(krate.add_version(try!(req.tx()), new_crate.vers.as_slice()));

    // Link this new version to all dependencies
    for dep in new_crate.deps.iter() {
        let tx = try!(req.tx());
        let krate = try!(Crate::find_by_name(tx, dep.name.as_slice()).map_err(|_| {
            human(format!("no known crate named `{}`", dep.name))
        }));
        try!(tx.execute("INSERT INTO version_dependencies \
                         (version_id, depends_on_id) VALUES ($1, $2)",
                        &[&vers.id, &krate.id]));
    }

    // Upload the crate to S3
    let handle = http::handle();
    let mut handle = match req.app().s3_proxy {
        Some(ref proxy) => handle.proxy(proxy.as_slice()),
        None => handle,
    };
    let path = krate.s3_path(new_crate.vers.as_slice());
    let (resp, cksum) = {
        let length = req.content_length().unwrap();
        let body = LimitErrorReader::new(req.body(), app.config.max_upload_size);
        let mut body = HashingReader::new(body);
        let resp = {
            let s3req = app.bucket.put(&mut handle, path.as_slice(), &mut body,
                                       "application/x-tar")
                                  .content_length(length)
                                  .header("Content-Encoding", "gzip");
            try!(s3req.exec().chain_error(|| {
                internal(format!("failed to upload to S3: `{}`", path))
            }))
        };
        (resp, body.final())
    };
    new_crate.cksum = cksum.as_slice().to_hex();
    if resp.get_code() != 200 {
        return Err(internal(format!("failed to get a 200 response from S3: {}",
                                    resp)))
    }

    // If the git commands fail below, we shouldn't keep the crate on the
    // server.
    struct Bomb { app: Arc<App>, path: Option<String>, handle: http::Handle }
    impl Drop for Bomb {
        fn drop(&mut self) {
            match self.path {
                Some(ref path) => {
                    let _ = self.app.bucket.delete(&mut self.handle,
                                                   path.as_slice())
                                .exec();
                }
                None => {}
            }
        }
    }
    let mut bomb = Bomb { app: app.clone(), path: Some(path), handle: handle };

    // Register this crate in our local git repo.
    let krate = try!(Crate::find_by_name(try!(req.tx()),
                                         new_crate.name.as_slice()));
    try!(git::add_crate(&**req.app(), &new_crate).chain_error(|| {
        internal(format!("could not add crate `{}` to the git repo", krate.name))
    }));

    // Now that we've come this far, we're committed!
    bomb.path = None;

    #[deriving(Encodable)]
    struct R { ok: bool, krate: EncodableCrate }
    Ok(req.json(&R { ok: true, krate: krate.encodable(Vec::new()) }))
}

fn parse_new_headers(req: &mut Request) -> CargoResult<(NewCrate, User)> {
    // Peel out all input parameters
    fn header<'a>(req: &'a Request, hdr: &str) -> CargoResult<Vec<&'a str>> {
        req.headers().find(hdr).require(|| {
            human(format!("missing header: {}", hdr))
        })
    }
    let auth = try!(header(req, "X-Cargo-Auth"))[0].to_string();
    let name = try!(header(req, "X-Cargo-Crate-Name"))[0].to_string();
    let vers = try!(header(req, "X-Cargo-Crate-Version"))[0].to_string();
    let feat = try!(header(req, "X-Cargo-Crate-Feature"))[0].to_string();
    let deps = try!(req.headers().find("X-Cargo-Crate-Dep").unwrap_or(Vec::new())
                       .iter().flat_map(|s| s.as_slice().split(';'))
                       .map(Dependency::parse)
                       .collect::<CargoResult<Vec<_>>>());
    let feat = match json::decode(feat.as_slice()) {
        Ok(map) => map,
        Err(..) => return Err(human("malformed feature header")),
    };

    // Make sure the tarball being uploaded looks sane
    let length = try!(req.content_length().require(|| {
        human("missing header: Content-Length")
    }));
    let max = req.app().config.max_upload_size;
    if length > max { return Err(human(format!("max upload size is: {}", max))) }
    {
        let ty = try!(header(req, "Content-Type"))[0];
        if ty != "application/x-tar" {
            return Err(human(format!("expected `application/x-tar`, \
                                      found `{}`", ty)))
        }
        let enc = try!(header(req, "Content-Encoding"))[0];
        if enc != "gzip" && enc != "x-gzip" {
            return Err(human(format!("expected `gzip`, found `{}`", enc)))
        }
    }

    // Make sure the api token is a valid api token
    let user = try!(User::find_by_api_token(try!(req.tx()),
                                            auth.as_slice()).map_err(|_| {
        human("invalid or unknown auth token supplied")
    }));

    // Validate the name parameter and such
    let new_crate = NewCrate {
        name: name.as_slice().chars().map(|c| c.to_lowercase()).collect(),
        vers: vers,
        deps: deps,
        features: feat,
        cksum: String::new(),
    };
    if !Crate::valid_name(new_crate.name.as_slice()) {
        return Err(human(format!("invalid crate name: `{}`", new_crate.name)))
    }
    if !Version::valid(new_crate.vers.as_slice()) {
        return Err(human(format!("invalid crate version: `{}`", new_crate.vers)))
    }
    Ok((new_crate, user))
}

pub fn download(req: &mut Request) -> CargoResult<Response> {
    let crate_name = req.params()["crate_id"].as_slice();
    let version = req.params()["version"].as_slice();
    let tx = try!(req.tx());
    let stmt = try!(tx.prepare("SELECT versions.id as version_id
                                FROM crates
                                LEFT JOIN versions ON
                                    crates.id = versions.crate_id
                                WHERE crates.name = $1
                                  AND versions.num = $2
                                LIMIT 1"));
    let mut rows = try!(stmt.query(&[&crate_name as &ToSql, &version as &ToSql]));
    let row = try!(rows.next().require(|| human("crate or version not found")));
    let version_id: i32 = row.get("version_id");
    let now = ::now();

    // Bump download counts.
    //
    // Note that this is *not* an atomic update, and that's somewhat
    // intentional. It doesn't appear that postgres supports an atomic update of
    // a counter, so we just do the hopefully "least racy" thing. This is
    // largely ok because these download counters are just that, counters. No
    // need to have super high-fidelity counter.
    //
    // Also, we only update the counter for *today*, nothing else. We have lots
    // of other counters, but they're all updated later on via the
    // update-downloads script.
    let amt = try!(tx.execute("UPDATE version_downloads
                               SET downloads = downloads + 1
                               WHERE version_id = $1 AND date($2) = date(date)",
                              &[&version_id, &now]));
    if amt == 0 {
        try!(tx.execute("INSERT INTO version_downloads
                         (version_id, downloads, counted, date, processed)
                         VALUES ($1, 1, 0, date($2), false)",
                        &[&version_id, &now]));
    }

    // Now that we've done our business, redirect to the actual data.
    let redirect_url = format!("https://{}/pkg/{}/{}-{}.tar.gz",
                               req.app().bucket.host(),
                               crate_name, crate_name, version);

    if req.wants_json() {
        #[deriving(Encodable)]
        struct R { ok: bool, url: String }
        Ok(req.json(&R{ ok: true, url: redirect_url }))
    } else {
        Ok(req.redirect(redirect_url))
    }
}
