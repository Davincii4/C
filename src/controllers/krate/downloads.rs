//! Endpoint for exposing crate download counts
//!
//! The enpoints for download a crate and exposing version specific
//! download counts are located in `krate::downloads`.

use std::cmp;

use crate::controllers::prelude::*;

use crate::models::{Crate, CrateVersions, Version, VersionDownload};
use crate::schema::version_downloads;
use crate::views::EncodableVersionDownload;

use crate::models::krate::to_char;

/// Handles the `GET /crates/:crate_id/downloads` route.
pub fn downloads(req: &mut dyn Request) -> AppResult<Response> {
    use diesel::dsl::*;
    use diesel::sql_types::BigInt;

    let crate_name = &req.params()["crate_id"];
    let conn = req.db_conn()?;
    let krate = Crate::by_name(crate_name).first::<Crate>(&*conn)?;

    let mut versions = krate.all_versions().load::<Version>(&*conn)?;
    versions.sort_by(|a, b| b.num.cmp(&a.num));
    let (latest_five, rest) = versions.split_at(cmp::min(5, versions.len()));

    let downloads = VersionDownload::belonging_to(latest_five)
        .filter(version_downloads::date.gt(date(now - 90.days())))
        .order(version_downloads::date.asc())
        .load(&*conn)?
        .into_iter()
        .map(VersionDownload::encodable)
        .collect::<Vec<_>>();

    let sum_downloads = sql::<BigInt>("SUM(version_downloads.downloads)");
    let extra = VersionDownload::belonging_to(rest)
        .select((
            to_char(version_downloads::date, "YYYY-MM-DD"),
            sum_downloads,
        ))
        .filter(version_downloads::date.gt(date(now - 90.days())))
        .group_by(version_downloads::date)
        .order(version_downloads::date.asc())
        .load::<ExtraDownload>(&*conn)?;

    #[derive(Serialize, Queryable)]
    struct ExtraDownload {
        date: String,
        downloads: i64,
    }
    #[derive(Serialize)]
    struct R {
        version_downloads: Vec<EncodableVersionDownload>,
        meta: Meta,
    }
    #[derive(Serialize)]
    struct Meta {
        extra_downloads: Vec<ExtraDownload>,
    }
    let meta = Meta {
        extra_downloads: extra,
    };
    Ok(req.json(&R {
        version_downloads: downloads,
        meta,
    }))
}

/// Handles the `GET /crates/:crate_id/recent_downloads` route.
pub fn recent_downloads(req: &mut dyn Request) -> AppResult<Response> {
    use diesel::dsl::*;

    let ndays = 90;

    let crate_name = &req.params()["crate_id"];
    let conn = req.db_conn()?;
    let krate = Crate::by_name(crate_name).first::<Crate>(&*conn)?;

    // Get the versions for this crate
    let versions = krate.all_versions().load::<Version>(&*conn)?;

    #[derive(Debug, Serialize)]
    struct Download<'a> {
        version: &'a semver::Version,
        downloads: i32,
    }

    #[derive(Debug, Serialize)]
    struct Response<'a> {
        downloads: Vec<Download<'a>>,
        meta: Meta<'a>,
    }

    #[derive(Debug, Serialize)]
    struct Meta<'a> {
        #[serde(rename = "crate")]
        krate: &'a str,
        ndays: i32,
    }

    // Now get the grouped versions for the last `ndays` days.
    //
    // XXX I am not sure how to do this in the database yet, with Diesel's API, so for the time
    // being, perform this aggregation in Rust.
    let downloads = VersionDownload::belonging_to(versions.as_slice())
        .filter(version_downloads::date.gt(date(now - ndays.days())))
        .load(&*conn)?
        .grouped_by(versions.as_slice())
        .into_iter()
        .map(|grouped_versions: Vec<VersionDownload>| {
            let total_downloads = grouped_versions.iter().map(|v| v.downloads).sum();
            // XXX(perf) this is slow, iterating over the versions array looking for the matching
            // version.
            let version = versions
                .iter()
                .find(|v| v.id == grouped_versions[0].version_id)
                .unwrap();

            Download {
                version: &version.num,
                downloads: total_downloads,
            }
        })
        .collect::<Vec<Download<'_>>>();

    Ok(req.json(&Response {
        downloads,
        meta: Meta {
            krate: crate_name,
            ndays,
        },
    }))
}
