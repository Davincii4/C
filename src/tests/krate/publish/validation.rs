use crate::builders::PublishBuilder;
use crate::util::{RequestHelper, TestApp};
use crates_io::models::krate::MAX_NAME_LENGTH;
use http::StatusCode;
use insta::assert_json_snapshot;

#[test]
fn invalid_names() {
    let (app, _, _, token) = TestApp::full().with_token();

    let bad_name = |name: &str| {
        let crate_to_publish = PublishBuilder::new(name, "1.0.0");
        let response = token.publish_crate(crate_to_publish);
        assert_eq!(response.status(), StatusCode::OK);
        assert_json_snapshot!(response.into_json());
    };

    bad_name("");
    bad_name("foo bar");
    bad_name(&"a".repeat(MAX_NAME_LENGTH + 1));
    bad_name("snow☃");
    bad_name("áccênts");

    bad_name("std");
    bad_name("STD");
    bad_name("compiler-rt");
    bad_name("compiler_rt");
    bad_name("coMpiLer_Rt");

    assert!(app.stored_files().is_empty());
}

#[test]
fn license_and_description_required() {
    let (app, _, _, token) = TestApp::full().with_token();

    let crate_to_publish = PublishBuilder::new("foo_metadata", "1.1.0")
        .unset_license()
        .unset_description();

    let response = token.publish_crate(crate_to_publish);
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.into_json());

    let crate_to_publish = PublishBuilder::new("foo_metadata", "1.1.0").unset_description();

    let response = token.publish_crate(crate_to_publish);
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.into_json());

    let crate_to_publish = PublishBuilder::new("foo_metadata", "1.1.0")
        .unset_license()
        .license_file("foo")
        .unset_description();

    let response = token.publish_crate(crate_to_publish);
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.into_json());

    assert!(app.stored_files().is_empty());
}
