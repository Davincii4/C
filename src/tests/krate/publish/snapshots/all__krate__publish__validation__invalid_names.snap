---
source: src/tests/krate/publish/validation.rs
expression: response.into_json()
---
{
  "errors": [
    {
      "detail": "the crate name cannot be an empty"
    }
  ]
}
